calicoctl/commands/ipam/check.go

```diff
@@ -0,0 +1,576 @@
+		}
+		numWEPIPs := 0
+		for _, w := range weps.Items {
+			ips, err := getWEPIPs(w)
+			if err != nil {
+				return err
+			}
+			for _, ip := range ips {
+				c.recordInUseIP(ip, w, fmt.Sprintf("Workload(%s/%s)", w.Namespace, w.Name))
+				numWEPIPs++
+			}
+		}
+		fmt.Printf("Found %d workload IPs.\n", numWEPIPs)
+		fmt.Printf("Workloads and nodes are using %d IPs.\n", len(c.inUseIPs))
+		fmt.Println()
+	}
+
+	{
+		const numNodesToPrint = 20
+		fmt.Printf("Looking for top (up to %d) nodes by allocations...\n", numNodesToPrint)
+		var allNodes []string
+		for n := range c.allocationsByNode {
+			allNodes = append(allNodes, n)
+		}
+		sort.Slice(allNodes, func(i, j int) bool {
+			// Reverse order
+			return len(c.allocationsByNode[allNodes[i]]) > len(c.allocationsByNode[allNodes[j]])
+		})
+		for i, n := range allNodes {
+			if i >= numNodesToPrint {
+				break
+			}
+			fmt.Printf("  %s has %d allocations\n", n, len(c.allocationsByNode[n]))
+		}
+		if len(allNodes) > 0 {
+			max := len(c.allocationsByNode[allNodes[0]])
+			median := len(c.allocationsByNode[allNodes[len(allNodes)/2]])
+			fmt.Printf("Node with most allocations has %d; median is %d\n", max, median)
+		}
+		fmt.Println()
+	}
+
+	numProblems := 0
+	var allocatedButNotInUseIPs []string
+	{
+		fmt.Printf("Scanning for IPs that are allocated but not actually in use...\n")
+		for ip, allocs := range c.allocations {
+			if _, ok := c.inUseIPs[ip]; !ok {
+				if c.showProblemIPs {
+					for _, alloc := range allocs {
+						fmt.Printf("  %s leaked; attrs %v\n", ip, alloc.GetAttrString())
+					}
+				}
+				allocatedButNotInUseIPs = append(allocatedButNotInUseIPs, ip)
+			}
+		}
+		numProblems += len(allocatedButNotInUseIPs)
+		fmt.Printf("Found %d IPs that are allocated in IPAM but not actually in use.\n", len(allocatedButNotInUseIPs))
+	}
+
+	var inUseButNotAllocatedIPs []string
+	var nonCalicoIPs []string
+	{
+		fmt.Printf("Scanning for IPs that are in use by a workload or node but not allocated in IPAM...\n")
+		for ip, owners := range c.inUseIPs {
+			if c.showProblemIPs && len(owners) > 1 {
+				fmt.Printf("  %s has multiple owners.\n", ip)
+			}
+			if _, ok := c.allocations[ip]; !ok {
+				// The IP is being used, but is not allocated within Calico IPAM!
+
+				// Found indicates whether the IP falls within an active IP pool.
+				found := false
+				parsedIP := net.ParseIP(ip)
+				for _, cidr := range activeIPPools {
+					if cidr.Contains(parsedIP) {
+						found = true
+						break
+					}
+				}
+				if !found {
+					if c.showProblemIPs {
+						for _, owner := range owners {
+							fmt.Printf("  %s in use by %v is not in any active IP pool.\n", ip, owner.FriendlyName)
+						}
+					}
+					nonCalicoIPs = append(nonCalicoIPs, ip)
+					continue
+				}
+				if c.showProblemIPs {
+					for _, owner := range owners {
+						fmt.Printf("  %s in use by %v and in active IPAM pool but has no IPAM allocation.\n", ip, owner.FriendlyName)
```

Comment:
Should we also write these into the report if "show the problem IPs" option is set?

Reply:
Yes, I think that we probably should. Although we can't actually do anything about these addresses if we find them - we just need to instruct the user to delete whatever thing is using the IP, most likely.

Reply:
I think this should be done in a follow-up, though. The main purpose of this is to get the initial "check and release" workflow functioning, which this isn't needed for.